tempesta_fw/http_parser.c

```diff
@@ -1375,6 +1395,338 @@ __parse_content_type(TfwHttpMsg *hm, unsigned char *data, size_t len)
 	return r;
 }
 
+static int
+__strdup_multipart_boundaries(TfwHttpReq *req)
+{
+	unsigned char *data_raw, *data, *ptr_raw, *ptr;
+	TfwStr *c, *end;
+
+	data_raw = tfw_pool_alloc(req->pool, req->multipart_boundary_raw.len);
+	data = tfw_pool_alloc(req->pool, req->multipart_boundary.len);
+
+	if (!data_raw || !data)
+		return -ENOMEM;
+
+	ptr_raw = data_raw;
+	ptr = data;
+	TFW_STR_FOR_EACH_CHUNK(c, &req->multipart_boundary_raw, end) {
+		memcpy_fast(ptr_raw, c->ptr, c->len);
+		ptr_raw += c->len;
+		if (c->flags & TFW_STR_VALUE) {
+			memcpy_fast(ptr, c->ptr, c->len);
+			ptr += c->len;
+		}
+	}
+
+	BUG_ON(ptr_raw != data_raw + req->multipart_boundary_raw.len);
+	BUG_ON(ptr != data + req->multipart_boundary.len);
+
+	req->multipart_boundary_raw.ptr = data_raw;
+	req->multipart_boundary.ptr = data;
+	__TFW_STR_CHUNKN_SET(&req->multipart_boundary_raw, 0);
+	__TFW_STR_CHUNKN_SET(&req->multipart_boundary, 0);
+
+	return 0;
+}
+
+static int
+__req_parse_content_type(TfwHttpMsg *hm, unsigned char *data, size_t len)
+{
+	int r = CSTR_NEQ;
+	TfwHttpReq *req = (TfwHttpReq *)hm;
+	__FSM_DECLARE_VARS(hm);
+
+	__FSM_START(parser->_i_st) {
+
+	__FSM_STATE(I_ContType) {
+		if (req->method != TFW_HTTP_METH_POST)
+			__FSM_I_JMP(I_EoL);
+		__FSM_I_JMP(I_ContTypeMediaType);
+	}
+
+	__FSM_STATE(I_ContTypeMediaType) {
+		static const TfwStr s_multipart_form_data =
+			TFW_STR_FROM("multipart/form-data");
+		TRY_STR_LAMBDA_fixup(&s_multipart_form_data, &parser->hdr, {},
+				     I_ContTypeMaybeMultipart);
+		if (chunk->len >= sizeof("multipart/") - 1) {
+			TRY_STR_INIT();
+			__FSM_I_JMP(I_ContTypeOtherSubtype);
+		} else {
+			TRY_STR_INIT();
+			__FSM_I_JMP(I_ContTypeOtherType);
+		}
+	}
+
+	__FSM_STATE(I_ContTypeMaybeMultipart) {
+		if (c == ';') {
+			set_bit(TFW_HTTP_F_CT_MULTIPART, req->flags);
```

Comment:
This is incorrect usage of `set_bit()` function. First parameter must be bit number, not the flag value. Correct expression is to be `set_bit(TFW_HTTP_B_CT_MULTIPART, req->flags)` and `TFW_HTTP_F_CT_MULTIPART` macro is not needed at all. You can't see errors, because the issue is happening at  every  `test_bit()/set_bit()` operation.

Reply:
Removed both `TFW_HTTP_F_CT_MULTIPART` and `TFW_HTTP_F_CT_MULTIPART_HAS_BOUNDARY`, and changed to `TFW_HTTP_B_*`.